api/envoy/extensions/filters/http/ratelimit/v3/rate_limit.proto

```diff
@@ -304,6 +304,18 @@ message RateLimitConfig {
       // [#extension-category: envoy.rate_limit_descriptors]
       config.core.v3.TypedExtensionConfig extension = 9;
     }
+
+    // If true, the action will be applied when the HTTP stream completes. The default value is false.
+    // This is useful when the rate limit action needs to be applied based on the response context.
+    //
+    // For example, let's say the upstream service calculates the usage statistics and returns them in the response,
+    // and we want to utilize these numbers to apply the rate limit action for the subsequent requests with this option.
+    // Combined with another filter that can set the descriptor value based on the response (e.g. Lua filter), this can be used
+    // to subtract the usage statistics from the rate limit descriptor.
+    //
+    // This is "lazy" by nature, and the rate limit won't be applied to the current request, but to the
+    // subsequent requests that match the descriptor.
+    bool apply_on_stream_done = 10;
```

Comment:
to be honest I am not sure `apply_on_stream_done` sounds correct naming - i am 100% open to renaming!